campaigns: implement licence checks

LawnGnome opened the pull request:
This PR implements the newer licensing behaviour discussed at sprint planning: campaigns may be used without a licence for campaigns with no more than five changesets, but attempting to create a campaign spec with more than that results in an error.

Closes #15715.

eseliger commented on enterprise/internal/campaigns/resolvers/errors.go:
```diff
@@ -30,6 +30,14 @@ func (e ErrCampaignsDisabledForUser) Extensions() map[string]interface{} {
+// to add an error code.
+type ErrCampaignsUnlicensed struct{ error }
+
+func (e ErrCampaignsUnlicensed) Extensions() map[string]interface{} {
+	return map[string]interface{}{"code": "ErrCampaignsUnlicensed"}
+}
```
❤️

eseliger commented on client/web/src/enterprise/campaigns/list/CampaignListIntro.tsx:
```diff
@@ -2,56 +2,73 @@ import React from 'react'
+            <div className="col-12 col-md-6 mb-2">
+                <DismissibleAlert className="campaign-list-intro__alert" partialStorageKey="campaign-list-intro-trial">
+                    <div className="campaign-list-intro__card card p-2 h-100">
+                        <div className="card-body d-flex align-items-start">
+                            {/* d-none d-sm-block ensure that we hide the icon on XS displays. */}
+                            <SourcegraphIcon className="mr-3 col-2 mt-2 d-none d-sm-block" />
+                            <div>
+                                <h4>Campaigns trial</h4>
+                                <p className="text-muted mb-0">
+                                    Campaigns will be a paid feature in a future release. In the meantime, we invite you
+                                    to trial the ability to make large scale changes across many repositories and code
+                                    hosts. If you’d like to discuss use cases and features,{' '}
+                                    <a href="https://about.sourcegraph.com/contact/sales/">please get in touch</a>!
+                                </p>
+                            </div>
                         </div>
                     </div>
-                </div>
-            </DismissibleAlert>
+                </DismissibleAlert>
```
I would probably extract this into a separate component like `CampaignChangelog`, because it's easier to read what _this_ component actually does.

LawnGnome replied:
Yep, I'm going to do this.

eseliger commented on client/web/src/enterprise/campaigns/list/CampaignListIntro.tsx:
```diff
@@ -2,56 +2,73 @@ import React from 'react'
+}
+
+export const CampaignListIntro: React.FunctionComponent<CampaignListIntroProps> = ({ licensed }) => {
+    if (licensed !== false) {
```
In the loading state it would flicker at first, if I'm not mistaken. I would do a 

```ts
if (licensed === undefined) {
return null
}
```

as well, to avoid that

LawnGnome replied:
I was kind of choosing my flicker there: not rendering the boxes while loading the licence state means that the rest of the content will move down once the licence state is available, which I personally hate. That said, it's not like this might not result in vertical movement as well.

With fresher eyes today, though, I might play around with having an explicit loading state over the top of the content with `visibility: hidden`, and then popping it in based on that. As long as the changelog is the same length or longer than the "you need a licence" wording, that would address that.

eseliger replied:
I mean, we could also defer the entire render until this AND the campaign are loaded, if that makes it easier for you :)

LawnGnome replied:
I don't think I know how to defer the render with the `FilteredConnection` in there, but I also don't think it's worthwhile. We should be able to optimise for the common case here (user has licence) easily enough.

LawnGnome replied:
I came up with a clever solution using a CSS grid to calculate the maximum height to avoid flickering, and then realised that it didn't play nicely if both `DismissableAlert`s were dismissed.

So I basically went with your idea instead: don't display anything except an empty row with the right margins, and then pop in the box(es) once we have the licensing state. This might encourage users to dismiss the boxes after the first time, but at least it's only one move down. (And only once per page load, given the `useMemo()`.)

eseliger commented on client/web/src/enterprise/campaigns/list/backend.ts:
```diff
@@ -151,3 +153,21 @@ export const queryCampaignsByNamespace = ({
+        `
+    ).pipe(
+        map(dataOrThrowErrors),
+        map(data => !!data.site?.productSubscription?.license?.tags.includes('campaigns'))
```
According to the GraphQL schema, this should be sufficient
```suggestion
        map(data => !!data.site.productSubscription.license?.tags.includes('campaigns'))
```

eseliger commented on client/web/src/enterprise/campaigns/list/CampaignListIntro.story.tsx:
```diff
@@ -0,0 +1,34 @@
+    </EnterpriseWebStory>
+))
+
+enum LicensingState {
+    Licensed = 'Licensed',
+    Unlicensed = 'Unlicensed',
+    Loading = 'Loading',
+}
+
+function stateToInput(state: LicensingState): boolean | undefined {
+    switch (state) {
+        case LicensingState.Licensed:
+            return true
+        case LicensingState.Unlicensed:
+            return false
+        default:
+            return undefined
+    }
+}
```
Cool!

eseliger replied:
If you want to be extra fancy, you could also make them additional top level stories, so the three cases are also run in CI for coverage :)

LawnGnome replied:
It's kind of unfortunate that we can't have the CI enumerate the available knobs, since these don't feel top-level-story-worthy, but I guess we are where we are. I'll add that.

eseliger reviewed (approved):
LGTM!

LawnGnome commented:
This has changed fairly significantly, so I'm re-requesting reviews. I also have a problem I can't figure out how to fix, so I'm waving the white flag in the hopes that someone (well, OK, probably @eseliger) knows what's going on.

#### Interesting changes since this morning

* I discovered that the licence check in the UI was insufficient: it checked for the licence tag, but tags can also be implied through licence plans. `licensing.Check()` handles this correctly on the backend, so rather than reimplementing that logic in TypeScript, I've added a new GraphQL query that essentially just delegates to that function. The catch is that it doesn't really fit in any of the existing enterprise resolvers, nor does it fit in our open source resolvers, so I've added a fifth enterprise-only resolver in `enterprise/internal/licensing` just to implement that query. There's a lot of boilerplate as a result.
* I realised that trying to avoid jumping content with dynamic `DismissableAlert` components was essentially impossible without pulling `DismissableAlert` apart and reimplementing it, so I've instead settled for making it so that there are zero or one jumps on page load: zero if the alert(s) have been dismissed, one if not. Since the licensing state is memoised after the first load, it shouldn't be too bad in practice.

#### Where I need help

I can't get the (many, _many_ expletives deleted) Puppeteer tests to pass. Additionally, I can't get them to run locally, which isn't exactly helping the cycle time on this. (I try something, wait 10 minutes for CI, and inevitably get distracted by something else I have left in this sprint.)

Fundamentally, the tests in `campaigns.test.ts` need a mock for the `AreCampaignsLicensed` query, which makes sense. I've added such a mock to the tests that seem to work for some of the tests in that file, but then some of them just don't seem to pick it up, even with the right incantation in place.

I don't know what I'm missing, but more importantly, I don't know how to debug this without being able to run them locally.

unknwon commented on enterprise/internal/campaigns/resolvers/resolver.go:
```diff
@@ -68,6 +69,27 @@ func campaignsCreateAccess(ctx context.Context) error {
+// checkLicense returns a user-facing error if the campaigns feature is not purchased
+// with the current license or any error occurred while validating the license.
+func checkLicense() error {
+	if !licensing.EnforceTiers {
```
From my impression, this feature flag won't be removed anytime soon. So does Campaigns want to enforce a valid license _right now_?

If so, you probably want to additionally call [`GetConfiguredProductLicenseInfo`](https://github.com/sourcegraph/sourcegraph/blob/dd7aff8ddb13b76c01a0f0d2591d5db3411972f5/enterprise/internal/licensing/licensing.go#L92) to check if a valid license is supplied (`err == nil`) inside this `if` branch.

mrnugget replied:
Thanks for this comment, @unknwon! @eseliger dug into this and this is what we found out:

Feature flag for licensing (`licensing.EnforceTiers`) is deactivated per default. That means current code (`770509c` in https://github.com/sourcegraph/sourcegraph/pull/16687) would’ve never enforced licensing.

We could skip checking the feature flag (by removing the `if !licensing.EnforceTiers` check here) and thus _enforce_ licensing, even if the feature flag is off.

That would mean:

- all existing customers with “old-enterprise” licenses would have campaigns enabled (that's what the code calls these licenses in `enterprise/internal/licensing/data.go`)
  - But **we don’t know who has old-enterprise licenses and who doesn’t**
  - we do have an old enterprise license in our dev environment
- only customers with “team” and “enterprise” licenses would be required to have the “campaigns” feature unlocked in their license

Options we have:

- Remove the feature-flag check to actually enforce licensing
- Change the code so that old-enterprise licenses don’t have access to campaigns

mrnugget reviewed (approved):
Code LGTM, but I think @unknwon uncovered a blocker.

mrnugget commented:
> I can't get the (many, _many_ expletives deleted) Puppeteer tests to pass. Additionally, I can't get them to run locally, which isn't exactly helping the cycle time on this. (I try something, wait 10 minutes for CI, and inevitably get distracted by something else I have left in this sprint.)

I'm looking into this now and maybe I can get it to work.

FWIW I got them to run locally by running the following in one terminal:

```
ENTERPRISE=1 yarn watch-web
```

and then in the other the failing test:

```
TS_NODE_PROJECT=client/web/src/integration/tsconfig.json NODE_NO_WARNINGS=1 yarn mocha --parallel=$CI --retries=1 ./client/web/src/integration/campaigns.test.ts -g 'displays a single campaign for org'
```

The specific command I got from the output of `yarn test-integration`. Info about how to run the integration tests: https://github.com/sourcegraph/sourcegraph/blob/dd7aff8ddb13b76c01a0f0d2591d5db3411972f5/doc/dev/background-information/testing.md#running-client-integration-tests

mrnugget commented:
I pushed a commit that fixes the failing puppeteer test. You missed one more mock where inside the test the mocks were reset.

So I guess, _I_ am the puppeteer now, huh 😎

LawnGnome reviewed (commented):
I can't approve my "own" PR, but the overnight changes LGTM. Thanks, Erik and Thorsten! :+1:
